pkg/activator/test/roundtripper.go

```diff
@@ -0,0 +1,145 @@
+*/
+
+package test
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"testing"
+
+	"github.com/knative/serving/pkg/network"
+	"github.com/knative/serving/pkg/queue"
+)
+
+// FakeResponse is a response given by the FakeRoundTripper
+type FakeResponse struct {
+	Err  error
+	Code int
+	Body string
+}
+
+// FakeRoundTripper is a roundtripper emulator useful in testing
+type FakeRoundTripper struct {
+	T *testing.T
+
+	// Return an error if host header does not match this
+	ExpectHost string
+
+	// lockerCh blocks responses being sent until a struct is written to the channel
+	LockerCh chan struct{}
+
+	// Responses to probe requests are popeed from this list until it is size 1 then
+	// that response is returned indefinitely
+	ProbeResponses []FakeResponse
+
+	// Response to non-probe requests
+	RequestResponse *FakeResponse
+
+	mux sync.Mutex
+}
+
+func defaultProbeResponse() *FakeResponse {
+	return &FakeResponse{
+		Err:  nil,
+		Code: 200,
+		Body: queue.Name,
+	}
+}
+
+func defaultRequestResponse() *FakeResponse {
+	return &FakeResponse{
+		Err:  nil,
+		Code: 200,
+		Body: "default response",
+	}
+}
+
+func response(fr *FakeResponse) (*http.Response, error) {
+	recorder := httptest.NewRecorder()
+	recorder.WriteHeader(fr.Code)
+	recorder.WriteString(fr.Body)
+	return recorder.Result(), nil
+}
+
+func popResponse(responses []FakeResponse, fallback *FakeResponse) (*FakeResponse, []FakeResponse) {
+	if responses == nil || len(responses) == 0 {
+		return fallback, responses
+	}
+	resp := &responses[0]
+	if len(responses) > 1 {
+		responses = responses[1:]
+	}
+	return resp, responses
+}
+
+func (rt *FakeRoundTripper) popResponse() *FakeResponse {
+	rt.mux.Lock()
+	defer rt.mux.Unlock()
+
+	resp, pr := popResponse(rt.ProbeResponses, defaultProbeResponse())
+	rt.ProbeResponses = pr
+	return resp
+}
+
+// RT is a RoundTripperFunc
+func (rt *FakeRoundTripper) RT(req *http.Request) (*http.Response, error) {
+	if req.Header.Get(network.ProbeHeaderName) != "" {
+		resp := rt.popResponse()
+		if resp.Err != nil {
+			return nil, resp.Err
+		}
+
+		// Make sure the probe is attributed with correct header.
+		if req.Header.Get(network.ProbeHeaderName) != queue.Name {
+			return response(&FakeResponse{
+				Code: http.StatusBadRequest,
+				Body: "probe sent to a wrong system",
+			})
+		}
+		return response(resp)
+	}
+	resp := rt.RequestResponse
+	if resp == nil {
+		resp = defaultRequestResponse()
+	}
+
+	if resp.Err != nil {
+		return nil, resp.Err
+	}
+
+	// verify that the request has the required rewritten host header.
+	if got, want := req.Host, ""; got != want {
+		return nil, fmt.Errorf("The req.Host has not been cleared out, was: %q", got)
+	}
+	if got, want := req.Header.Get("Host"), ""; got != want {
+		return nil, fmt.Errorf("The Host header has not been cleared out, was: %q", got)
+	}
+
+	if rt.ExpectHost != "" {
+		if got, want := req.Header.Get(network.OriginalHostHeader), rt.ExpectHost; got != want {
```

Comment:
Collapse the if's using `&&` maybe?

Reply:
The logic for that is slightly different in that we will have to always get the header, even when we don't want to assert on it.